File: content/15-greedy/jump_game.js
/**
 * Problem: Jump Game
 * LeetCode: #55 (Medium) | NeetCode 150 | Blind 75
 */

function canJump(nums) {
    /**
     * Time Complexity: O(N)
     * Space Complexity: O(1)
     */
    let goal = nums.length - 1;
    
    for (let i = nums.length - 2; i >= 0; i--) {
        if (i + nums[i] >= goal) {
            goal = i;
        }
    }
    
    return goal === 0;
}

if (require.main === module) {
    const assert = require('assert');
    assert.strictEqual(canJump([2, 3, 1, 1, 4]), true);
    assert.strictEqual(canJump([3, 2, 1, 0, 4]), false);
    console.log("JS: canJump passed tests!");
}

File: content/15-greedy/jump_game.py
"""
Problem: Jump Game
LeetCode: #55 (Medium) | NeetCode 150 | Blind 75
"""

def can_jump(nums: list[int]) -> bool:
    """
    Time Complexity: O(N)
    Space Complexity: O(1)
    """
    goal = len(nums) - 1
    
    for i in range(len(nums) - 2, -1, -1):
        if i + nums[i] >= goal:
            goal = i
            
    return goal == 0

if __name__ == "__main__":
    assert can_jump([2, 3, 1, 1, 4]) == True
    assert can_jump([3, 2, 1, 0, 4]) == False
    print("Python: can_jump passed tests!")
